Extension of Max Flow algorithms: undirected graphs are now supported (retry)

jkinable opened the pull request:
This pull request extends both EdmondsKarpMaximumFlow and PushRelabelMaximumFlow, thereby adding the option to run both algorithms on undirected graphs.

Changelog:
-EdmondsKarpMaximumFlow and PushRelabelMaximimumFlow algorithms now also work on undirected graphs
-Added test for both directed and undirected graphs
-Rewrote & documented significant parts of the old implementation. 
-Deleted RandomizedTest from Max flow tests: this test did not serve any purpose. The results were never verified. The only way the test could fail is when the algorithm would throw some internal exception.
-Deleted the Legacy implementation of EdmondsKarpMaximumFlow and corresponding test
-Documented public (!) functions of Extension class. Added some questions wrt the revision of this class.
-Separated flow calculation from actually generating a map representing the flow on each edge individually. The latter often is expensive and unnecessary.

Note: in the previous release of jgrapht, significant changes to the MaxFlow algorithms were introduced. Unfortunately, this code was overly complex, messy, and very poorly documented, especially when you consider the simplicity of the algorithm. It took a considerable amount of time to reverse engineer what was going on. A simple 1 hour task of adding support for undirected graphs turned out to be a 2-day endeavor. Let's try to avoid this in the future. Algorithms should be implemented (and documented) as clear as possible such that other people can extend the code, fix potential bugs, etc. In addition, the runtime complexity of an implementation should be clear.

Note 2: I left several comments in the Extension.java class. Please review those.

jsichi commented:
@alexeykudinkin :  would you mind reviewing this?

@jkinable : one man's "overly complex, messy, and very poorly documented" is another man's "elegantly self-evident".  For the future, let's avoid opinionated/contentious language in pull requests, and focus on the details of the code.  Thanks!

alexeykudinkin commented on jgrapht-core/src/test/java/org/jgrapht/alg/flow/EdmondsKarpMaximumFlowTest.java:
```diff
@@ -35,27 +35,30 @@
         DirectedWeightedMultigraph<Integer, DefaultWeightedEdge> simple =
-            new DirectedWeightedMultigraph<Integer, DefaultWeightedEdge>(
-                DefaultWeightedEdge.class);
+            new DirectedWeightedMultigraph<>(DefaultWeightedEdge.class);
```
It's better to abstract such a (notational) changes from any other (meaningful) changes. 
That'd greatly facilitate reviewing process.

jkinable replied:
good point, will make a separate pull request for that next time

alexeykudinkin commented on jgrapht-core/src/main/java/org/jgrapht/alg/flow/EdmondsKarpMaximumFlow.java:
```diff
@@ -144,64 +141,67 @@ public EdmondsKarpMaximumFlow(DirectedGraph<V, E> network, double epsilon)
             }
 
-            augmentFlow();
+            maxFlowValue+=augmentFlow();
```
Let's respect established formatting rules.

jkinable replied:
I've no clue what you mean here. It seems correctly formatted, and clearly readable to me. Plus, in terms of the algorithm it also makes sense since every augmenting flow increases the max flow.

alexeykudinkin replied:
Please, check out formatting of the other assignments in the file.

jkinable replied:
I honestly can't spot the difference compared to what you for example wrote: "reverseEdge.capacity += flowDifference"
I can't hardly imagine you are referring to the space before and after the +=?

alexeykudinkin replied:
Yes, it refers to spaces also.

alexeykudinkin commented on jgrapht-core/src/main/java/org/jgrapht/alg/flow/EdmondsKarpMaximumFlow.java:
```diff
@@ -59,20 +59,29 @@
 
     private final ExtensionFactory<VertexExtension> vertexExtensionsFactory;
-    private final ExtensionFactory<EdgeExtension> edgeExtensionsFactory;
+    private final ExtensionFactory<AnnotatedFlowEdge> edgeExtensionsFactory;
```
What are the reasons to change naming for the edge-extension? Why did vertex-extension's name stay untouched in this case?

jkinable replied:
In the original code there was an Extension, which was implemented or extended by a base extension which was then extended by ExtendedEdge. Then there were functions such as protected EdgeExtension extendedEdge(E e){return edgeExtended(e)}.There was a lot of something-extends-something which made this simple algorith rather complex. As you've noticed, function mentioned above has been eliminated and code has been simplified. If you know a catchy name for VertexExtension then I'm happy to replace that name as well.

alexeykudinkin commented on jgrapht-core/src/main/java/org/jgrapht/alg/flow/MaximumFlowAlgorithmBase.java:
```diff
@@ -60,58 +61,85 @@
      * Default tolerance.
      */
     public static final double DEFAULT_EPSILON = 1e-9;
+    public final double EPSILON; // tolerance (DEFAULT_EPSILON or user-defined)
```
CAPS naming-style is only used for the constant-fields, and is prohibited for the ordinary ones.

alexeykudinkin commented on jgrapht-core/src/main/java/org/jgrapht/alg/flow/EdmondsKarpMaximumFlow.java:
```diff
@@ -144,64 +141,67 @@ public EdmondsKarpMaximumFlow(DirectedGraph<V, E> network, double epsilon)
         if (!network.containsVertex(source)) {
             throw new IllegalArgumentException(
-                "invalid source (null or not from this network)");
+                    "invalid source (null or not from this network)");
```
Reasons for such a change? Let's try to mitigate such an irrelevant changes.

jkinable replied:
No clue how this sneaked in. At least not on purpose. Sorry about that.

alexeykudinkin commented on jgrapht-core/src/main/java/org/jgrapht/alg/flow/EdmondsKarpMaximumFlow.java:
```diff
@@ -258,8 +263,10 @@ private void augmentFlow()
                     seen))
             {
                 pushFlowThrough(ex, deltaFlow);
+                flowIncrease+=deltaFlow;
```
Style

alexeykudinkin commented on jgrapht-core/src/main/java/org/jgrapht/alg/flow/MaximumFlowAlgorithmBase.java:
```diff
@@ -60,58 +61,85 @@
 
-    <VE extends VertexExtensionBase, EE extends EdgeExtensionBase> void init(
+    protected Extension<V, ? extends VertexExtensionBase> vertexExtensionManager;
+    protected Extension<E, AnnotatedFlowEdge> edgeExtensionManager;
```
Reasons?

`edgeExtensionManager` has been made of type `Extension<E, ? extends EdgeExtensionBase>` for a purpose: to allow inheritors to have edge-extensions of their own. I can hardly spot any benefits for such a change.

jkinable replied:
I've changed this to `protected ExtensionManager<E, ? extends AnnotatedFlowEdge> edgeExtensionManager;`

alexeykudinkin commented on jgrapht-core/src/main/java/org/jgrapht/alg/flow/MaximumFlowAlgorithmBase.java:
```diff
@@ -60,58 +61,85 @@
+    protected Extension<E, AnnotatedFlowEdge> edgeExtensionManager;
+
+    protected double maxFlowValue=-1; //Max flow established after last invocation of the algorithm.
+    protected Map<E, Double> maxFlow=null; //Mapping of the flow on each edge.
```
Style

jkinable replied:
Not sure what makes you unhappy here. Perhaps you prefer /* comment */ better. I'll add that.

alexeykudinkin replied:
Formatting style abuse is what ultimately makes me unhappy about that snippet. Please, checkout the overall style of the code, and make your changes consistent with that.

I'm totally fine with EOL comments (and indeed prefer those to multiline) ones.

alexeykudinkin commented on jgrapht-core/src/main/java/org/jgrapht/alg/flow/MaximumFlowAlgorithmBase.java:
```diff
@@ -60,58 +61,85 @@
-    private Extension<V, ? extends VertexExtensionBase> vXs;
-    private Extension<E, ? extends EdgeExtensionBase> eXs;
+    protected Graph<V, E> network; //input network
+    protected final boolean DIRECTED_GRAPH; //indicates whether the input graph is directed or not
```
Same concerns as for an `EPSILON`.

alexeykudinkin commented on jgrapht-core/src/main/java/org/jgrapht/alg/flow/MaximumFlowAlgorithmBase.java:
```diff
@@ -60,58 +61,85 @@
+    public MaximumFlowAlgorithmBase(Graph<V,E> network, double epsilon){
+        this.network=network;
+        this.DIRECTED_GRAPH=network instanceof DirectedGraph;
+        this.EPSILON=epsilon;
```
Style

jkinable replied:
seems fine to me.

alexeykudinkin commented on jgrapht-core/src/main/java/org/jgrapht/alg/flow/MaximumFlowAlgorithmBase.java:
```diff
@@ -60,58 +61,85 @@
 
         buildInternal();
+        maxFlowValue=0;
+        maxFlow=null;
```
Style

jkinable replied:
seems fine to me

alexeykudinkin commented on jgrapht-core/src/main/java/org/jgrapht/alg/flow/MaximumFlowAlgorithmBase.java:
```diff
@@ -60,58 +61,85 @@
-
-        for (V u : n.vertexSet()) {
-            VertexExtensionBase ux = extendedVertex(u);
+        if(DIRECTED_GRAPH) { //Directed graph
```
Style

alexeykudinkin commented on jgrapht-core/src/main/java/org/jgrapht/alg/flow/MaximumFlowAlgorithmBase.java:
```diff
@@ -60,58 +61,85 @@
+                AnnotatedFlowEdge backwardEdge = createBackwardEdge(forwardEdge);
+                ux.getOutgoing().add(forwardEdge);
+                vx.getOutgoing().add(backwardEdge);
+            }
```
Style

alexeykudinkin commented on jgrapht-core/src/main/java/org/jgrapht/alg/flow/MaximumFlowAlgorithmBase.java:
```diff
@@ -120,118 +148,111 @@ private EdgeExtensionBase createEdge(
+    protected void pushFlowThrough(AnnotatedFlowEdge edge, double flow)
     {
-        EdgeExtensionBase iex = ex.<EdgeExtensionBase>getInverse();
+        AnnotatedFlowEdge reverseEdge = edge.getInverse();
```
Let's use consistent naming throughout the change.

jkinable replied:
corrected in 2 different places

alexeykudinkin commented on jgrapht-core/src/main/java/org/jgrapht/alg/flow/MaximumFlowAlgorithmBase.java:
```diff
@@ -120,118 +148,111 @@ private EdgeExtensionBase createEdge(
+     * @param flow
+     * @param val
+     * @return
+     */
```
Invalid doc.

alexeykudinkin commented on jgrapht-core/src/main/java/org/jgrapht/alg/flow/MaximumFlowAlgorithmBase.java:
```diff
@@ -120,118 +148,111 @@ private EdgeExtensionBase createEdge(
-        E prototype;
+        private VertexExtensionBase source; //Edge source
+        private VertexExtensionBase target; //Edge target
+        private AnnotatedFlowEdge inverse; //Inverse edge
```
Comments style

jkinable replied:
Left in from original code, but changed to /* style */

alexeykudinkin commented on jgrapht-core/src/main/java/org/jgrapht/alg/flow/MaximumFlowAlgorithmBase.java:
```diff
@@ -253,16 +274,67 @@ public void setTarget(VertexExtensionBase target)
+     * @return <i>read-only</i> mapping from edges to doubles - flow values
+     */
+    public Map<E, Double> getMaximumFlow(){
+        if(maxFlow==null) //Lazily calculate the max flow map
```
Style

alexeykudinkin commented on jgrapht-core/src/main/java/org/jgrapht/alg/flow/MaximumFlowAlgorithmBase.java:
```diff
@@ -253,16 +274,67 @@ public void setTarget(VertexExtensionBase target)
+            return annotatedFlowEdge.getTarget().prototype;
+        else
+            return reverseFlowEdge.getTarget().prototype;
+    }
```
Style

alexeykudinkin commented on jgrapht-core/src/main/java/org/jgrapht/alg/flow/PushRelabelMaximumFlow.java:
```diff
@@ -300,36 +300,34 @@ protected void pushFlowThrough(EdgeExtension ex, double f)
+//    private VertexExtension extendedVertex(V v)
+//    {
+//        return this.vertexExtended(v);
+//    }
```
Let's remove commented out things.

alexeykudinkin commented on jgrapht-core/src/main/java/org/jgrapht/alg/util/Extension.java:
```diff
@@ -34,60 +34,127 @@
+     * JK: Renamed this function to 'getInstance' for consistency with the getSingletonInstance function
+     * @return Extension object
+     */
+    @Deprecated
```
First of all, let's stay consistent: previous naming scheme wasn't ideal, but current pick is even worse -- `getInstance` is a widespread convention for naming singleton-getters, therefore seeing `edgeExtensionManager.getInstance` hurts readability as hard as it could.

Much better choice would be to name those pair `createExtension`/`getExtension`.

alexeykudinkin commented on jgrapht-core/src/main/java/org/jgrapht/alg/util/Extension.java:
```diff
@@ -34,60 +34,127 @@
+     * @param t prototype
+     * @return Extension of prototype
+     */
+    @Deprecated
```
See above

alexeykudinkin commented on jgrapht-core/src/main/java/org/jgrapht/alg/util/Extension.java:
```diff
@@ -34,60 +34,127 @@
+     * NOTE: JK - Why do we want or need this? The only function this would offer is that all extensions have some common
+     * super type. However, none of the code requires that E extends BaseExtension. Furthermore, I'm not sure why this
+     * class is a subclass of the Extension/ExtensionManager class, instead of a class on its own.
+     */
```
Yeah, missing comment may drive anyone nuts guessing why this thing is around. It serves the simple purpose to enforce extension to be default-constructible. And, yes, there're missing generic constraint for such a thing in `Extension`.

jkinable replied:
Not sure how to fix this nicely though. `public class Extension<T, E extends Extension.BaseExtension>` doesn't look very appealing, especially not since 'E' is used as a generic Edge throughout most of jgrapht. Suggestions/preferences?

alexeykudinkin replied:
It's perfectly fine serving its own role.

alexeykudinkin commented on jgrapht-core/src/main/java/org/jgrapht/alg/util/Extension.java:
```diff
@@ -34,60 +34,127 @@
-     * @param <E> extension concept class-type
+     * @param <E> class-type of extension
+     *
+     * NOTE: JK - Why is this class a subclass of the Extension/ExtensionManager class, instead of a class on its own.
```
To encapsulate things.

jkinable replied:
But if you create a 
```
public interface ExtensionFactory<E extends BaseExtension>
    {
        E create();
    }
then this could have been an external class right?
```

alexeykudinkin replied:
Don't see any proposal below?

alexeykudinkin commented on jgrapht-core/src/main/java/org/jgrapht/alg/util/Extension.java:
```diff
@@ -34,60 +34,127 @@
+     * Comments/Description missing
+     *
+     * NOTE: JK - Do we really need this? This seems well beyond what any user would require. Better to keep the library as concise as possible.
+     */
```
That thing has been exposed to manifest problems with construction of inheritors of the `ExtensionBase`. Though, since it stay unused, it could be killed safely.

jkinable commented:
hi @alexeykudinkin ,
Thanks again for the thorough revision. Apologies for my rather harsh comment at the beginning. I'm sure you've had the best intentions in mind when implementing/revising these classes. In particular, I think that your addition of a performance test class based on the Java Microbenchmark Harness is a great idea and a very good extension to the existing code!
Based on your/my comments, I would propose the following revision to the current 'Extension' class.

Revision proposal for the class Extension.
1. Rename the current Extension class to 'ExtensionManager'
2. Introduce a new class AbstractExtension (consistent to AbstractGraph).
3. ExtensionManager has functions "createExtension" and "getExtension" which return an <E extends AbstractExtension> object
4. Introduce a new class ExtensionFactory which creates <E extends AbstractExtension> objects
5. Remove the 'BaseExtension, ExtensionManagerInstantiationException, ExtensionFactory subclasses from the current Extension class

What do you think?

alexeykudinkin commented:
Hey, @jkinable!

That's ok, just let's keep in mind, that we're all determined to make this library both versatile, performant & consistent.

Proposed changes (to the `Extension` class) sound great to me. Though i've some outstanding concerns about other changes introduced.

jkinable commented:
@alexeykudinkin I've submitted a new version; all the issues you've raised have been addressed.

@jsichi:  Alexey and I revised the 'Extension' classes (see package org.jgrapht.alg.util.extension) in the last commit. Is this what you had in mind when you wrote that 'JgraphT needs a strongly-typed data structure parallel to the original graph structure, which provides a way to directly augment vertices/edges with additional data attributes'? Or are there things missing? Personally I'm not entirely sure what the added value of these classes is. If you want to add attributes to an edge, one could for example simply create a new "AnnotatedWeightedEdge extends DefaultWeightedEdge" class. Similar for vertices. The graph library already contains factory classes for vertices/edges. Even the implementation of the Flow algorithms which currently uses the Extension classes could be rewritten using the existing edge/vertex classes without loosing any functionality or extendibility. Either way, if you see a need for these classes I'm happy to include/augment/revise them.

alexeykudinkin commented on jgrapht-core/src/main/java/org/jgrapht/alg/flow/MaximumFlowAlgorithmBase.java:
```diff
@@ -118,13 +126,15 @@ private void buildInternal()
-            for(V v : network.vertexSet()) {
-                VertexExtensionBase vx = vertexExtensionManager.getSingletonInstance(v);
-                vx.prototype=v;
+            for(V v : network.vertexSet())
```
Let's convert `for(` to `for (`.

alexeykudinkin commented on jgrapht-core/src/main/java/org/jgrapht/alg/flow/MaximumFlowAlgorithmBase.java:
```diff
@@ -118,13 +126,15 @@ private void buildInternal()
-            for(E e : network.edgeSet()){
-                VertexExtensionBase ux=vertexExtensionManager.getSingletonInstance(network.getEdgeSource(e));
-                VertexExtensionBase vx=vertexExtensionManager.getSingletonInstance(network.getEdgeTarget(e));
+            for(E e : network.edgeSet())
```
Same here.

alexeykudinkin commented on jgrapht-core/src/main/java/org/jgrapht/alg/flow/MaximumFlowAlgorithmBase.java:
```diff
@@ -154,15 +164,15 @@ private AnnotatedFlowEdge createBackwardEdge(
-            if (!DIRECTED_GRAPH) { //undirected graph: if (u,v) exists, then so much (v,u)
-                backwardEdge.capacity= network.getEdgeWeight(backwardPrototype);
-                backwardEdge.prototype=backwardPrototype;
+            if (!directed_graph) { //undirected graph: if (u,v) exists, then so much (v,u)
```
Comment style, should follow the rule: `// Undirected graph bla-bla-bla`

alexeykudinkin commented:
Awesome re-touch @jkinable!

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/util/extension/AbstractExtension.java:
```diff
@@ -0,0 +1,11 @@
+package org.jgrapht.alg.util.extension;
```
Need file header

jsichi commented:
@jkinable:  Regarding the extension concept, the idea is that we don't want to duplicate the entire graph structure; we just want to augment it with additional information.  For example, if a particular algorithm needs to keep track of a color for each vertex, it shouldn't need to duplicate all the vertices and edges; it should just be able to keep its own (strongly-typed) information as annotations on the original graph (with a way of easily navigating back and forth).  For this purpose, algorithm-specific extension classes are the most appropriate solution (pace Occam).  Does that help make the purpose clearer?

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/util/Extension.java:
```diff
@@ -20,7 +20,7 @@
  * the Eclipse Foundation.
  */
 /* -----------------
- * MaximumFlowAlgorithmBase.java
```
I'm confused...shouldn't this file Extension.java be deleted?  Why is it still present in the pull request?

jkinable replied:
This file should indeed be deleted. But since it is present in the previous version, I made it deprecated such that it can be deleted in the next version. I'm fine with deleting it right-away though since I don't think that there will be many people who adapted to the extension concept since the last version.

jsichi replied:
Got it.  Let's delete it now.  1.0.0 is already going to be a breaking release where we get rid of all old deprecated stuff, so in this case, let's skip the deprecation and just get rid of it to avoid confusion.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/util/extension/AbstractExtension.java:
```diff
@@ -0,0 +1,45 @@
+ * An object, from here on denoted as prototype,can be encapsulated in or extended by another object. An example would be the relation between an edge (prototype) and an annotated edge. The
+ * annotated edge encapsulates/extends an edge, thereby augmenting it with additional data. In symbolic form, if b is the prototype class, than a(b) would be its extension. This concept is similar to java's extension where
+ * one class is derived from (extends) another class (prototype).
+ */
```
Any reason not to make this an interface instead?

jsichi replied:
(Actually I'm not entirely clear on the purpose of this...I didn't understand the comment from @alexeykudinkin about enforcing default-constructibility.)

jkinable replied:
I left it identical to the implementation currently present in jgrapht. It indeed seems more appropriate to make it an abstract class.

jsichi replied:
@alexeykudinkin is the idea of having it as an abstract class that we might eventually need to add data members here?

alexeykudinkin replied:
@jsichi no, that primarily was an occasion. It could be safely transformed into interface. 

P.S. Please, ignore my previous comment about default-constructibility.

jsichi commented:
@jkinable @alexeykudinkin :  Regarding terminology, I think Alexey's "prototype" language is a bit confusing.  Generally, the term "prototype" means something different in software:  a  prototype is an object instance which can be cloned in order to produce new instances.

https://en.wikipedia.org/wiki/Prototype-based_programming

My suggestion would be either "original" or "primary".

jsichi commented:
@jkinable @alexeykudinkin : Regarding style, up until now, I have run the Jalopy code formatter over all of jgrapht-core as part of each release, ensuring a consistent look (and allowing me to keep nitpicking to a minimum during code reviews).  Unfortunately, I'm not sure this is going to continue as a possibility for future releases now that we've moved to Java 8, since Jalopy is quite ancient.  (Maybe we can find a modern, open-source formatter as a replacement.)

Anyway, as much as possible, sticking to the existing formatting (even with respect to trivial things such as spaces and braces) is appreciated.

And definitely for naming conventions, which formatters can't fix automatically (e.g. we never use lower_case_with_undercore, so directed_graph should be renamed to isDirectedGraph).

alexeykudinkin commented on jgrapht-core/src/main/java/org/jgrapht/alg/flow/MaximumFlowAlgorithmBase.java:
```diff
@@ -60,58 +62,94 @@
 
-    <VE extends VertexExtensionBase, EE extends EdgeExtensionBase> void init(
+    protected ExtensionManager<V, ? extends VertexExtensionBase> vertexExtensionManager;
+    protected ExtensionManager<E, ? extends AnnotatedFlowEdge> edgeExtensionManager;
```
I still don't get it: why its better than `EdgeExtension`? That sole change blows the whole change-set up, while carrying around 0 profit.

jkinable commented:
I've reworked your previous comments:
-Deleted the old Extension class
-Changed the word 'prototype' to 'original'
-Changed the class 'AbstractExtension from an abstract class to an interface. Furthermore, since the class is no longer 'abstract', it seemed appropriate to rename it to 'Extension'. Since the 'Extension' class is in a different package as the old 'Extension' class, this change should not break any old implementations'

I hope that I've hereby addressed all issues and that this extensive set of changes can be merged into the main branch somewhere in the foreseeable future. I'm unfortunately running out of time to make additional changes. Thanks again @alexeykudinkin for your valuable comments.

jsichi commented:
Thanks Joris!  There are still a bunch of references to "prototype", but let's put this one to bed for now and address that in a followup.
